Using REQUEST buckets in HTTP processing

icing opened the pull request:
## Abstract

 * core: expect the new REQUEST buckets from input filters when reading the next request on a connection. No longer parses itself or is tied to a specific HTTP version format.
 * mod_http: install a new HTTP1_REQUEST_IN input filter that reads the HTTP/1.x lines from the network and transforms it into a REQUEST bucket for handling in the core.

## Changes

The HTTP/1.x parsing code in `server/protocol.c` has moved to `modules/http/http_filter.c` and `modules/http/http_protocol.c`. The checking on HTTP conformance (seen in the old `rrl_error` enum) has been split into the generic `http_error` enum in core and the rest has been moved to mod_http and the parsing code.

This means that `ap_read_request()`, the work horse in connection processing, now is agnostic to exactly *how* a request came into the server. The task of transforming network bytes to HTTP requests is now the task of the input filters. Core only expects either a REQUEST or an ERROR bucket. Everything else is a general error for aborting the connection. ERROR buckets can be used to specify the status code for the HTTP response.

On connections, where `http/1.1` is the protocol, mod_http installs the new input filter HTTP1_REQUEST_IN that basically does the parts formerly in `ap_read_request()`. 

## Motivation

As with the PRs regarding HEADERS and RESPONSE buckets, making the httpd core agnostic to specific protocol versions avoids code duplication and workarounds in non HTTP/1.x handling. This comes at basically no cost for HTTP/1.x processing  as the same code is run, only in a slightly different sequence.

## TBD

Again, the changes to mod_http2 are not part of this PR to make review easier. There are several parts that can be removed after this PR. Also, h2 worker threads will no longer really be h2 specific. Which means we could think about a "secondary connection workers" pool in mpm. And monitoring/scoreboards for that etc. etc.

Commit:
*) core/mod_http: use REQUEST meta buckets and a new HTTP/1.x specific
     input filter to separate the handling for HTTP requests from the
     handling of HTTP/1.x request parsing and checks.
     A new HTTP1_REQUEST_IN filter installs itself on http/1.1 connections
     before a request is being read. It generates either a REQUEST meta
     bucket on success or an ERROR bucket with the proposed response status.
     The core connection processing, relying on ap_read_request(), now expects
     a REQUEST or ERROR bucket from the input filters and is agnostic to
     specific HTTP versions and how they bring requests into the server.

icing commented on include/mod_core.h:
```diff
@@ -54,6 +55,10 @@ apr_status_t ap_http_filter(ap_filter_t *f, apr_bucket_brigade *b,
+apr_status_t ap_h1_request_in_filter(ap_filter_t *f, apr_bucket_brigade *bb,
+                                     ap_input_mode_t mode, apr_read_type_e block,
+                                     apr_off_t readbytes);
+
```
The new input filter that does the parsing of network bytes.

icing commented on modules/http/http_core.c:
```diff
@@ -269,15 +271,73 @@ static int http_create_request(request_rec *r)
+        if (r->proxyreq == PROXYREQ_NONE) {
+            ap_add_input_filter_handle(ap_h1_request_in_filter_handle,
+                                       NULL, r, r->connection);
+        }
```
The filter installs itself on non-proxy http/1.1 connections.

icing commented on modules/http/http_core.c:
```diff
@@ -269,15 +271,73 @@ static int http_create_request(request_rec *r)
+                                   NULL, r, r->connection);
+    }
+    return OK;
+}
```
Install the body filter very early, after a request has been read so that anyone added later can rely on its handling.

icing commented on modules/http/http_core.c:
```diff
@@ -269,15 +271,73 @@ static int http_create_request(request_rec *r)
+    }
+    return OK;
+}
+
```
Very late, after all post hook have run, we check HTTP/1.x specific request properties, e.g. transfer-encoding and the need to close the connection after this request.

ylavic replied:
Why this late and not in h1_post_read_request_early() too?
A post_read_request hook or input filter run from there might need to know about TE vs CL no?

icing replied:
Valid point. I was thinking about hooks that might tinker with the request regarding this. But you are right, hooks might want to access that information.

I'll move it.

icing replied:
done.

icing replied:
done

icing commented on modules/http/http_filters.c:
```diff
@@ -1856,12 +1856,12 @@ AP_CORE_DECLARE_NONSTD(apr_status_t) ap_h1_response_out_filter(ap_filter_t *f,
 
                 ap_log_rerror(APLOG_MARK, APLOG_TRACE2, 0, r,
-                              "ap_http1_response_out_filter seeing response bucket status=%d",
+                              "ap_h1_response_out_filter seeing response bucket status=%d",
```
Just a rename forgotten in the previous PR.

icing commented on modules/http/http_filters.c:
```diff
@@ -2080,3 +2080,184 @@ static apr_bucket *create_trailers_bucket(request_rec *r, apr_bucket_alloc_t *bu
+    char *uri;
+    char *protocol;
+} h1_request_ctx;
+
```
The input filter context that collects all request information to generate the REQUEST bucket.

Code is basically moved here from `server/protocol.c`.

icing commented on modules/http/http_protocol.c:
```diff
@@ -1636,3 +1636,195 @@ AP_DECLARE(void) ap_h1_add_end_chunk(apr_bucket_brigade *b,
+    rrl_none, rrl_badprotocol, rrl_badmethod, rrl_badwhitespace, rrl_excesswhitespace,
+    rrl_missinguri, rrl_baduri, rrl_trailingtext,
+} rrl_error;
+
```
Error checks formerly in `server/protocol.c` that are HTTP/1.x specific and happen during request parsing.

icing commented on server/protocol.c:
```diff
@@ -682,359 +682,13 @@ static int field_name_len(const char *field)
     return end - field;
 }
 
-static const char m_invalid_str[] = "-";
```
All this has moved to `modules/http/`.

ylavic replied:
I'm not sure it needed to, it makes the changes harder to follow and protocol.c seems to be calling ap_h1_ specific functions still.
Maybe move them after the functional changes if needed?

icing commented on server/protocol.c:
```diff
@@ -682,359 +682,13 @@ static int field_name_len(const char *field)
-    return 1;
-}
-
 AP_DECLARE(int) ap_parse_request_line(request_rec *r)
```
This function was in the API before and I left it here, but it is now using code from `modules/http`.

ylavic replied:
Yeah all the story of the protocol.c, hard to move things to a module..

icing commented on server/protocol.c:
```diff
@@ -1466,22 +1120,196 @@ static void apply_server_config(request_rec *r)
+    http_error_badprotocol,
+    http_error_reject09,
+    http_error_badmethod09,
+} http_error;
```
The parts from the former `rrl_error` that apply to HTTP in general. Used when assigning a REQUEST bucket to a `request_rec`.

icing commented on server/protocol.c:
```diff
@@ -1466,22 +1120,196 @@ static void apply_server_config(request_rec *r)
+    r->proto_num = proto_num;
+    return error;
+}
+
```
Code formerly part of HTTP/1.x parsing that now checks the values delivered in a REQUEST bucket.

icing commented on server/protocol.c:
```diff
@@ -1466,22 +1120,196 @@ static void apply_server_config(request_rec *r)
-    (void)method;
-    (void)uri;
-    (void)protocol;
+    core_server_config *conf = ap_get_core_module_config(r->server->module_config);
```
Assign `method`, `uri` and `protocol` to a `request_rec`. These are coming from a received REQUEST bucket and need to be checked for validity. And if valid, all concerned fields need to be updated.

Again, code moved here from former http/1.x parsing.

ylavic replied:
Maybe call it ap_assign_request_line()?

icing replied:
🚴‍♀️Makes it easier to understand for all the h1 boomers, yes.

icing replied:
done.

icing commented on server/protocol.c:
```diff
@@ -1490,8 +1318,59 @@ request_rec *ap_read_request(conn_rec *conn)
-    /* Get the request... */
-    if (!read_request_line(r, tmp_bb) || !ap_parse_request_line(r)) {
+    r->request_time = apr_time_now();
+    rv = ap_get_brigade(r->proto_input_filters, tmp_bb, AP_MODE_READBYTES, APR_BLOCK_READ, 0);
```
This is a bit quirky,  to read 0 bytes from the input filter to get only the meta buckets. But, if we read any more, we need to push it back somehow so that the request handlers can read it all during processing.

icing commented on server/protocol.c:
```diff
@@ -1490,8 +1318,59 @@ request_rec *ap_read_request(conn_rec *conn)
+        else if (AP_BUCKET_IS_ERROR(e)) {
+            access_status = ((ap_bucket_error *)(e->data))->status;
+            goto die_unusable_input;
+        }
```
The filter detected an error that should result in a HTTP response with the error status code.

ylavic replied:
Ah OK, so the input error bucket is caught here, why not.
So up to ap_read_request() some special input filtering is in place and after that only body filtering right?
Could you please show the output of `dump_filters r->[proto_]input_filters` (from our gdbinit) at different ap_read_request() steps? This'd possibly help have a picture here, or however you'd like to describe it :)

icing replied:
Added a macro in util_filter.h to dump a filter chain and now see in the logs for a http/1.1 request:

```
filter-dump before pre_read[0]: reqtimeout [58 CONNECTION]
filter-dump before pre_read[1]: log_input_output [59 CONNECTION]
filter-dump before pre_read[2]: core_in [60 NETWORK]
filter-dump after pre_read[0]: http1_request_in [30 PROTOCOL]
filter-dump after pre_read[1]: reqtimeout [58 CONNECTION]
filter-dump after pre_read[2]: log_input_output [59 CONNECTION]
filter-dump after pre_read[3]: core_in [60 NETWORK]
ap_tokenize_request_line: 'GET / HTTP/1.1'
ap_tokenize_request: error=0, method=GET, uri=/, protocol=HTTP/1.1
http1 request and headers parsed: GET / HTTP/1.1
filter-dump after get_brigade[0]: reqtimeout [58 CONNECTION]
filter-dump after get_brigade[1]: log_input_output [59 CONNECTION]
filter-dump after get_brigade[2]: core_in [60 NETWORK]
checking request: GET / HTTP/1.1
assigned protocol: HTTP/1.1, error=0
assigned request_line: GET / HTTP/1.1, error=0
parsed uri: r->status=200, error=0
filter-dump after post_read[0]: http_in [30 PROTOCOL]
filter-dump after post_read[1]: http1_body_in [40 TRANSCODE]
filter-dump after post_read[2]: reqtimeout [58 CONNECTION]
filter-dump after post_read[3]: log_input_output [59 CONNECTION]
filter-dump after post_read[4]: core_in [60 NETWORK]
```

icing commented on server/protocol.c:
```diff
@@ -1490,8 +1318,59 @@ request_rec *ap_read_request(conn_rec *conn)
+        }
+        else if (!APR_BUCKET_IS_METADATA(e) && e->length != 0) {
+            break;
+        }
```
We should not really get those on our 0 length read, but better be safe.

ylavic replied:
So shouldn't we error out in this case (i.e. even if we saw REQUEST bucket)?

icing replied:
I think we should indeed. It will not work if we get request body buckets before we have read and analysed the request. Will add the check and fail.

icing replied:
done.

icing commented on server/protocol.c:
```diff
@@ -1490,8 +1318,59 @@ request_rec *ap_read_request(conn_rec *conn)
+        uri = apr_pstrdup(r->pool, breq->uri);
+        protocol = apr_pstrdup(r->pool, breq->protocol);
+        headers = breq->headers? apr_table_clone(r->pool, breq->headers) : NULL;
+    }
```
Some optimisation on assigning the values. If the input filter used `r->pool` for the values in the REQUEST bucket, we can just work with these. Otherwise copy.

ylavic commented on modules/http/http_filters.c:
```diff
@@ -2080,3 +2080,184 @@ static apr_bucket *create_trailers_bucket(request_rec *r, apr_bucket_alloc_t *bu
+    }
+
+    /* This filter needs to get out of the way of read_request_line() */
+    ap_remove_input_filter(f);
```
Hm, enough for now I suppose but it does not look nonblocking compatible. We probably should think of nonblocking in the new design..

icing replied:
Agreed. I wanted to limit the scope of this PR in just moving pieces around. And the current `ap_read_request` is blocking. But we could make it non-blocking with run_connection returning on a APR_EAGAIN.
Is that what you have in mind?

ylavic replied:
That's what I meant indeed, this filter will have to be reworked to not remove itself unconditionally on the first call then.

icing replied:
Yes, I'd like `ap_read_request()` to get the brigade NONBLOCK.

ylavic commented on modules/http/http_filters.c:
```diff
@@ -2080,3 +2080,184 @@ static apr_bucket *create_trailers_bucket(request_rec *r, apr_bucket_alloc_t *bu
+            }
+            /* got the request line and it looked to contain what we need */
+            ctx->state = REQ_HEADERS;
+            break;
```
fall through?

ylavic commented on modules/http/http_filters.c:
```diff
@@ -2080,3 +2080,184 @@ static apr_bucket *create_trailers_bucket(request_rec *r, apr_bucket_alloc_t *bu
+
+        case REQ_BODY:
+            /* we should not come here */
+            AP_DEBUG_ASSERT(0);
```
It seems to me that REQ_BODY and REQ_ERROR are hardly expected here, let's not code them and have a single `default:` case which assert()s + EINVAL?

icing replied:
I wanted to modify/improve the existing code as little as possible, so the review can be more clear. Maybe we'll have a look at it in another PR afterwards for code improvements?

ylavic replied:
+1

ylavic commented on modules/http/http_filters.c:
```diff
@@ -2080,3 +2080,184 @@ static apr_bucket *create_trailers_bucket(request_rec *r, apr_bucket_alloc_t *bu
+cleanup:
+    if (http_status != HTTP_OK) {
+        ap_log_rerror(APLOG_MARK, APLOG_TRACE3, 0, r,
+                      "failed reading request line, returning error bucket %d", http_status);
```
Returning an error bucket from input filters in not usual (and probably not expected by upstream filters and handlers, why not return rv and let the caller decide with ap_map_http_request_error()?

[edit: I saw it was caught in ap_read_request() actually? so early enough for later handling to never see an error bucket?]

icing replied:
Then a filter that knows the HTTP status it wants to trigger would have to reverse engineer ap_map_http_request_error()'s mapping? It seemed rather straightforward to use ERROR buckets for this, but maybe I fail to see the complications.

ylavic replied:
I was wondering if any handler calling ap_get_brigade(), like mod_proxy forwarding or even the default_handler(), could now receive a bucket as error signal (while they usually don't handle input errors like this) thus simply ignoring the error.
Then I saw ap_read_request() which handles the error bucket and [edit]ed my comment, IIUC it shouldn't be an issue because ap_h1_request_in_filter() is supposed to remove itself before ap_read_request() ends.

ylavic commented on modules/http/http_protocol.c:
```diff
@@ -1636,3 +1636,195 @@ AP_DECLARE(void) ap_h1_add_end_chunk(apr_bucket_brigade *b,
+    return (len > LOG_NAME_MAX_LEN)? LOG_NAME_MAX_LEN : (int)len;
+}
+
+static void rrl_log_error(request_rec *r, rrl_error error, char *etoken)
```
**const** char *etoken

icing replied:
Will change.

ylavic commented on modules/http/http_protocol.c:
```diff
@@ -1636,3 +1636,195 @@ AP_DECLARE(void) ap_h1_add_end_chunk(apr_bucket_brigade *b,
+
+static void rrl_log_error(request_rec *r, rrl_error error, char *etoken)
+{
+    if (error == rrl_badprotocol)
```
switch (error)?

icing replied:
Will change.

icing replied:
done

ylavic commented on modules/http/http_protocol.c:
```diff
@@ -1636,3 +1636,195 @@ AP_DECLARE(void) ap_h1_add_end_chunk(apr_bucket_brigade *b,
+
+AP_DECLARE(int) ap_h1_tokenize_request_line(
+        request_rec *r, const char *line,
+        char **pmethod, char **puri, char **pprotocol)
```
The returned `char**`s seem to be passed to ap_assign_request() and probably don't need to be modified inplace by the caller, so `const char**`? This would apply to tokenize_request_line() then too.

icing replied:
done.

ylavic commented on server/protocol.c:
```diff
@@ -1531,18 +1410,7 @@ request_rec *ap_read_request(conn_rec *conn)
-        }
-
-        clen = apr_table_get(r->headers_in, "Content-Length");
+        const char *clen = apr_table_get(r->headers_in, "Content-Length");
```
IIUC, we should have accurate r->body_indeterminate at this point, so check !r->body_indeterminate here and fail/internal error if no C-L in this case?

icing replied:
I believe this is too early. Currently, we set this in the post_read hook. Also, requests may have neither C-L or body_indeterminate, e.g. if they are known to have no request body.

Maybe converting body_indeterminate into a no_body flag would make things more clear?

Commit:
Updates after first review by ylavic:
- rename ap_assign_request to ap_assign_request_line
- added some const goodness to params for tokenizing request line
- checking for non-0 data buckets in ap_read_request() and failing
  if detected.

Commit:
update after first review by ylavic:
- have a single post_read_request hook for http/1.x that runs
  early and sets body_indeterminate + installs the body filter.

ylavic reviewed (approved):
+1

icing commented:
Thanks for reviewing! Merged as r1899799 into trunk.
